File: mag_task/ur_main.py
# Write your solution here

import time
import math
from ur_aux import VrepModel

# image processing
import numpy as np
import cv2 

#################### Functions ###########################

# Put value into range 0 - 255
def rng(x):
  return max(min(x,255),0)

# Find pegs in the image
# Return: [(x1,y1,rad1),(x2,y2,rad2)...]  
def findPegs(image, accumulator=10):
  _,gray = cv2.threshold(image[:,:,2], 180, 255, cv2.THRESH_TOZERO)  # pegs are red
  # find circles
  circles = cv2.HoughCircles(gray, cv2.HOUGH_GRADIENT, 1, 50, param1=50, param2=accumulator, minRadius=10, maxRadius=40)
  if circles is None:
    return []
  else:
    circles = np.round(circles[0,:]).astype('int')    
    isRed = lambda p,q: image[q,p,0] < 100 and image[q,p,1] < 100 and image[q,p,2] > 180
    return [x for x in circles if isRed(rng(x[0]),rng(x[1]))]
   

# Show image and circles
# image - image array
# lst - list of found circles, could be empty
def visualize(image,lst=[]):
  img = image.copy()
  if lst:
    for (x,y,r) in lst:
      cv2.circle(img, (x,y), r, (0,255,0), 4)
      cv2.rectangle(img, (x-5,y-5), (x+5,y+5), (0,128,255), -1)      
  cv2.imshow('view', img)
  cv2.waitKey(0)

##################### Prepare ###############################

# port is in remoteApiConnections.txt
port = 19997
# simplify function call
rad = math.radians


##################### Start #############################

ur = VrepModel(port)   # create object
ur.startSimulation()   # start to work

#################### Solution ###########################

# Write your solution here
# You can use the following functions or write your own.

ur.ptp([0,rad(20),rad(-75),rad(-35),rad(90),0])     # go to some position in joint space

pos = ur.getToolPosition() # read current position

pos[1] -= 0.1
ur.lin(pos)            # go to new position in Cartesian space

ur.gripperOpen(True)   # use True/False to open/close the gripper

img = ur.getImage()    # get camera image
pegs = findPegs(img)   # try to find pegs

visualize(img,pegs)    # you can use it to visualize the picture, press any key to close the window


#################### Finish ###############################
ur.ptp([0,0,0,0,0,0])   # go to base 
ur.check()              # check the result
ur.stopSimulation()

File: mag_task/task_cpp/ur_main.cpp
// Write your code here

#include "py_bind.h"
#include <opencv2/opencv.hpp>

#define RAD(X) (3.1415926*X/180)

int main(int argc, char** argv) 
{
  /************ Start simulation ******************/ 
  UR ur(19997);          // connection port number
  ur.startSimulation();

  /********* Write your solution *****************/
  
  // point to point motion
  double q0[] = {0,RAD(20),RAD(-75),RAD(-35),RAD(90),0};
  ur.ptp(q0);

  double q1[JOINT_NO];
  ur.getJointPosition(q1);
  // check
  std::cout << q1[1] << " " << q1[3] << " " << q1[5] << std::endl;

  // tool 
  ur.gripperOpen(true);

  // cartesian motion
  double p[CART_NO];
  double o[CART_NO];

  ur.getToolPosition(p);
  ur.getToolOrientation(o);
  // check
  std::cout << "Position " << p[0] << " " << p[1] << " " << p[2] << std::endl;
  std::cout << "Orientation " << o[0] << " " << o[1] << " " << o[2] << std::endl;

  p[1] -= 0.1;
  p[0] += 0.1;
  ur.lin(p);

  o[2] -= 0.1;
  ur.lin(p,o);
  
  // read force and torque
  double force[CART_NO];
  double torque[CART_NO];
  ur.getForceTorque(force,torque);
  // check
  std::cout << force[0] << " " << torque[0] << std::endl;
  
  // get image
  ImgData data;
  ur.getImage(&data);
  cv::Mat img = cv::Mat(data.height, data.width, CV_8UC3, data.buffer, cv::Mat::AUTO_STEP);
  cv::imshow("image",img);
  cv::waitKey(0);

  /******************** Finish ***************************/
  double qend[] = {0,0,0,0,0,0};
  ur.ptp(qend);
  ur.check();
  ur.stopSimulation();
  
  return 0;
}
